File: src/Jobs/MiningJob.cpp
//
//  MiningJob.cpp
//  SpaceStationKeeper
//
//  Created by Sean Dawson on 28/04/2014.
//
//

#include "MapPoint.h"
#include "GameGrid.h"
#include "Resource.h"
#include "Job.h"
#include "MiningJob.h"
#include "GameTile.h"

MiningJob::MiningJob(GameGrid *gameGrid, MapPoint targetMapPoint,
                     AssetLoaderBase *assetLoader)
    : _gameGrid(gameGrid), _targetMapPoint(targetMapPoint),
      _assetLoader(assetLoader) {};
MiningJob::~MiningJob() {};

bool MiningJob::preRequisitesAreMet() {
  GameTile *tile = _gameGrid->getTile(_targetMapPoint);
  string assetRef = tile->material->texture->getAssetRef();
  cout << "MiningJob::preRequisitesAreMet(): assetRef: " << assetRef
       << " tile addr: " << tile << endl;
  return tile != nullptr && assetRef == string(MiningJob::preReqTileSet);
}
bool MiningJob::postRequistesAreMet() {
  GameTile *tile = _gameGrid->getTile(_targetMapPoint);
  string assetRef = tile->material->texture->getAssetRef();
  return tile != nullptr && assetRef == string(MiningJob::postReqTileSet);
}
bool MiningJob::isDone() { return _workUnitsDone >= MiningJob::maxWorkUnits; }

float MiningJob::getProgress() {
  return _workUnitsDone / (float)MiningJob::maxWorkUnits;
}
void MiningJob::update(float deltaTime) {
  float workPerSecond = MiningJob::workUnitPerWorker * 2;
  float workRate = workPerSecond * deltaTime;
  _workUnitProgress += workRate;

  if (_workUnitProgress >= 1.0f) {
    _workUnitsDone++;
    _workUnitProgress = 0.0f;
  }

  if (isDone()) {
    GameTile *tile = _gameGrid->getTile(_targetMapPoint);
    MaterialRef materialCopy = MaterialRef(new Material{*tile->material});
    TextureDefRef corridorTd =
        _assetLoader->loadAsset<TextureDef>(MiningJob::postReqTileSet);
    materialCopy->texture = corridorTd;
    tile->material = materialCopy;
    tile->material->baseColor = ColorAf(0.0f, 0.0f, 1.0f, 1.0f);
  }
}

File: assets/scripts/jobs/MiningJob/MiningJob.lua
-- Mining Job
class 'MiningJob' (Job)

function MiningJob:__init(gameGrid, coord, preReqTexture, postReqTexture)
    Job.__init(self)
    LuaDebug.Log('MiningJob:__init() called!')
    self._gameGrid = gameGrid
    self._coord = coord
    self._preReqTexture = preReqTexture
    self._postReqTexture = postReqTexture
    self._workUnitsDone = 0
    self._maxWorkUnits = 8
    self._workUnitPerWorker = 1 --TODO: put in config
    self._workUnitProgress = 0
end

function MiningJob:preRequisitesAreMet()
    local tile = self._gameGrid.tiles[self._coord.x..self._coord.y]
    local tex = tile.renderer.material.texture
    return tile ~= nil and tex == self._preReqTexture
end

function MiningJob:postRequistesAreMet()
    local tile = self._gameGrid.tiles[self._coord.x..self._coord.y]
    local tex = tile.renderer.material.texture
    return tile ~= nil and tex == self._postReqTexture
end

function MiningJob:isDone()
    return self._workUnitsDone >= self._maxWorkUnits
end

function MiningJob:getProgress()
    return self._workUnitsDone / self._maxWorkUnits
end

function MiningJob:update(deltaTime)
    local workPerSecond = self._workUnitPerWorker * self:getWorkerCount() 
    local workRate = workPerSecond * deltaTime
    self._workUnitProgress = self._workUnitProgress + workRate

    if (self._workUnitProgress >= 1) then
        self._workUnitsDone = self._workUnitsDone + 1
        self._workUnitProgress = 0
    end

    if (self:isDone()) then
        local tile = self._gameGrid.tiles[self._coord.x..self._coord.y]
        local sprite = GetComponentFromType(tile, 'Sprite')
        sprite:setSpriteTexture(self._postReqTexture)
        --sprite:setup()
        self._gameGrid:FixTileFrames(self._coord)
        GetBotManager():addCoord(self._coord)
    end
end

function MiningJob:getStartLocation() 
    return self._coord
end

function MiningJob:getEndLocation()
    return self._coord
end

function MiningJob:getRadius() 
    return 1 -- Mining happens on the adjacent tile not on the actual tile that is inaccessible
end

function MiningJob:getMaxWorkers()
    return 2
end

function MiningJob:getWorkerSlot(workerIndex) 
    if workerIndex == 0 then
        return Vec2f(-0.7, 0.7)
    end
    if workerIndex == 1 then
        return Vec2f(0.7, 0.7)
    end
    error("Invalid worker index: Should be between 0 and " .. tostring(self:getMaxWorkers() - 1))

end